Sale Coupon Multi-Currency

oerp-odoo opened the pull request:
[IMP] sale_coupon_multi_use: multi-currency support
    
    Multi Use coupons are now multi currency aware and will convert used
    amounts back to program currency.
    
    [ADD] sale_coupon_multi_currency: init
    
    Makes currency field visible on coupon/promotion program.
    
    Can set custom currency instead of just using company currency.

oerp-odoo committed:
[ADD] sale_coupon_multi_currency: init

Makes currency field visible on coupon/promotion program.

Can set custom currency instead of just using company currency.

oerp-odoo committed:
[IMP] sale_coupon_multi_use: multi-currency support

Multi Use coupons are now multi currency aware and will convert used
amounts back to program currency.

i-vyshnevska commented on sale_coupon_multi_currency/models/sale_coupon.py:
```diff
@@ -0,0 +1,9 @@
+from odoo import fields, models
```
```suggestion
# Copyright 2020 Camptocamp SA
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
from odoo import fields, models
```

i-vyshnevska commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -57,9 +59,22 @@ def _get_related_sale_order_line(self, sale_order):
     def _prepare_consumption_line(self, sale_order_line):
         self.ensure_one()
+        program = self.program_id
+        from_currency = sale_order_line.order_id.pricelist_id.currency_id
```
```suggestion
        from_currency = sale_order_line.order_id.currency_id
```

oerp-odoo replied:
Can't find source, but I remember it was said that explicit full path to record is better performance wise instead of using related field, because you don't need to initiate related field logic.

i-vyshnevska replied:
my point here that currency on sale order may differ from price list currency because of customer settings or manual change etc.

oerp-odoo replied:
How can it differ? Its related field that points to pricelist currency:

```python
    currency_id = fields.Many2one("res.currency", related='pricelist_id.currency_id', string="Currency", readonly=True, required=True)
```

i-vyshnevska replied:
oh, sorry than I missed some, I thought it was optional

jcoux commented on sale_coupon_multi_currency/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,27 @@
+
+    _inherit = "sale.coupon.program"
+
+    currency_id = fields.Many2one(
+        "res.currency",
+        related=None,
+        compute="_compute_currency_id",
+        inverse="_inverse_currency_id",
+    )
```
```suggestion
    currency_id = fields.Many2one(
        "res.currency",
        string="Currency",
        compute="_compute_currency_id",
        inverse="_inverse_currency_id",
    )
```

jcoux commented on sale_coupon_multi_currency/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,27 @@
+    @api.depends("company_id.currency_id", "currency_custom_id")
+    def _compute_currency_id(self):
+        for rec in self:
+            rec.currency_id = rec.currency_custom_id.id or rec.company_id.currency_id.id
```
```suggestion
            rec.currency_id = rec.currency_custom_id or rec.company_id.currency_id
```

jcoux commented on sale_coupon_multi_currency/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,27 @@
+
+    def _inverse_currency_id(self):
+        for rec in self:
+            rec.currency_custom_id = rec.currency_id.id
```
```suggestion
            rec.currency_custom_id = rec.currency_id
```

jcoux commented on sale_coupon_multi_currency/tests/common.py:
```diff
@@ -0,0 +1,26 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+
```
```suggestion
```

jcoux replied:
```suggestion
```

jcoux commented on sale_coupon_multi_currency/tests/test_sale_coupon_multi_currency.py:
```diff
@@ -0,0 +1,30 @@
+        )
+        self.assertEqual(self.company_main.currency_id, self.currency_company)
+        # Case 2.
+        self.program_coupon_percentage.currency_custom_id = False
```
`currency_custom_id` is normally not modified directly, but using inverse function on `currency_id`:
```suggestion
        self.program_coupon_percentage.currency_id = False
```

oerp-odoo replied:
Well this is test. It won't work, because then it unsets company currency. This is a test case where we look what would be computed if custom currency is unset. From interface, you can't unset that currency, so this is a bit different than mimicking `currency_id` inverse.

jcoux commented on sale_coupon_multi_currency/views/sale_coupon_program_views.xml:
```diff
@@ -0,0 +1,21 @@
+            <field name="company_id" position="before">
+                <field
+                    name="currency_id"
+                    readonly="False"
```
No need readonly flag as you have an inverse function on the field.
```suggestion
```

oerp-odoo replied:
Yes you do. If I dont set it, field becomes readonly.

jcoux commented on sale_coupon_multi_currency/views/sale_coupon_program_views.xml:
```diff
@@ -0,0 +1,21 @@
+                <field
+                    name="currency_id"
+                    readonly="False"
+                    required="True"
```
If `currency_id` is required for all `sale.coupon.program`, I would prefer to have the required directly on the field definition in python. No?

oerp-odoo replied:
`currency_id` is computed and not stored, so don't know if it makes sense to mark it as readonly on python. Usually it means, required on database level.

i-vyshnevska replied:
computed and related already readonly fields

oerp-odoo replied:
Will such approach won't have problems on existing records, once it is installed? I guess default will be populated there?

jcoux commented on sale_coupon_multi_currency/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,27 @@
+
+    def _inverse_currency_id(self):
+        for rec in self:
+            rec.currency_custom_id = rec.currency_id.id
```
I have doubt about this.
As the `currency_id` is required on the view and as it's the only field displayed on the view, finally, `currency_custom_id` will be always set.
In this case, I think it's better to have only one field `currency_id` and with a default value with the company currency.

And with only one field, the code will be more easy, we will have lower fields, and the maintainability will be better.

oerp-odoo replied:
@jcoux 
Hm, and how will we know which company record is related with program when trying to use that company default currency?
When you call default method, its not record bound, so it does not have any info regarding specific record, thus it won't know which company its related at that point. Of course its possible to just use user company, but it won't be correct, cause user company might not match program company.

This issue is also related with default value on already existing program records (when module will be installed).

Then I might need to write post init hook, to set default value after module was installed.

Wouldn't that make it more complicated?

jcoux replied:
As discussed directly, I didn't realized that currency_id already exists in code core.
Then your way with 2 fields seems good :+1: 
Just please add a comment to be explicit.

jcoux commented on sale_coupon_multi_currency/views/sale_coupon_views.xml:
```diff
@@ -0,0 +1,13 @@
+        <field name="inherit_id" ref="sale_coupon.sale_coupon_view_form" />
+        <field name="arch" type="xml">
+            <field name="code" position="before">
+                <field name="currency_program_id" invisible="1" />
```
Why don't display it as information?
And as it's a related field, it will be readonly.

oerp-odoo replied:
because its only useful if you are using it for something. I added it for reusability. Whats the point to show it there, if its not gonna be related with any amount fields?

jcoux commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -65,10 +70,9 @@ msgstr "Montant restant"
-"Fixed Amount can't be changed when there are Multi Use coupons already."
+"Fixed Amount or Custom Currency can't be changed when there are Multi Use "
+"coupons already."
 msgstr ""
```
```suggestion
msgstr ""
"Le montant restant ne peut pas être modifié quand un bon de réduction a déjà"	
" été utilisé plusieurs fois."
```

jcoux commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -127,9 +131,14 @@ msgstr "Bons de réduction"
 #: model:ir.model,name:sale_coupon_multi_use.model_sale_coupon_apply_code
 msgid "Sales Coupon Apply Code"
-msgstr "Code du bon de réduction"
+msgstr "Codice di applicazione coupon di vendita"
```
```suggestion
msgstr "Code du bon de réduction"
```

jcoux commented on sale_coupon_multi_use/__manifest__.py:
```diff
@@ -8,7 +8,7 @@
     "author": "Camptocamp, Odoo Community Association (OCA)",
     "license": "AGPL-3",
-    "depends": ["sale_coupon"],
+    "depends": ["sale_coupon_multi_currency"],
```
You have a problem because with your implementation, we can't install `sale_coupon_multi_use` without `sale_coupon_multi_currency`.
Modules must not have dependencies between them.
We must can have a multi use of coupons without multi currencies and the inverse too.

Don't finish the review on multi_use module as you will redo this part.

oerp-odoo replied:
I was thinking about that, but thought that won't make a difference if these are separated. If you want I can glue em.

jcoux reviewed (changes requested):
You have a problem because with your implementation, we can't install `sale_coupon_multi_use` without `sale_coupon_multi_currency`.
Modules must not have dependencies between them.
We must can have a multi use of coupons without multi currencies and the inverse too.

Don't finish the review on multi_use module as you will redo this part.

N.B: If you **really** need parts with dependencies on the 2 modules, you need to do a glue module auto-installed with both modules in dependencies and which contains only parts that required the 2 modules.
But the 2 modules must be installed separately.

oerp-odoo committed:
[REF] sale_coupon_multi|use|currency

Decoupled dependency between sale_coupon_multi_use and
sale_coupon_multi_currency, so it could be used independently, but also
added bridge module to make sure currency can't be changed if multi
use coupon is already used.

Also fixed currency conversion bug from coupon to sale order amount -
conversion must have been done before comparing SO amount not after,
because amounts compared could be in different currencies.

jcoux commented on sale_coupon_multi_currency/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,29 @@
+    @api.depends("company_id.currency_id", "currency_custom_id")
+    def _compute_currency_id(self):
+        for rec in self:
+            rec.currency_id = rec.currency_custom_id.id or rec.company_id.currency_id.id
```
```suggestion
            rec.currency_id = rec.currency_custom_id or rec.company_id.currency_id
```

jcoux commented on sale_coupon_multi_currency/models/sale_coupon_program.py:
```diff
@@ -0,0 +1,29 @@
+
+    def _inverse_currency_id(self):
+        for rec in self:
+            rec.currency_custom_id = rec.currency_id.id
```
```suggestion
            rec.currency_custom_id = rec.currency_id
```

jcoux commented on sale_coupon_multi_currency/tests/test_sale_coupon_multi_currency.py:
```diff
@@ -0,0 +1,30 @@
+        )
+        self.assertEqual(self.company_main.currency_id, self.currency_company)
+        # Case 2.
+        self.program_coupon_percentage.currency_custom_id = False
```
To have a behavior like in interface:
```suggestion
        self.program_coupon_percentage.currency_id = False
        self.program_coupon_percentage._inverse_currency_id()
```

oerp-odoo replied:
I already answered you this why it wont work on test. And there is no `currency_id =  False` on interface, because field is required.

jcoux replied:
Ok, then forgot this comment.

jcoux commented on sale_coupon_multi_use/i18n/fr.po:
```diff
@@ -127,9 +133,14 @@ msgstr "Bons de réduction"
 #: model:ir.model,name:sale_coupon_multi_use.model_sale_coupon_apply_code
 msgid "Sales Coupon Apply Code"
-msgstr "Code du bon de réduction"
+msgstr "Codice di applicazione coupon di vendita"
```
```suggestion
msgstr "Code du bon de réduction"
```

jcoux commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -19,11 +19,14 @@ class SaleCoupon(models.Model):
     # Takes value from related program (coupon_multi_use field), when
     # it is generated.
     multi_use = fields.Boolean(readonly=True)
+    currency_program_id = fields.Many2one(related="program_id.currency_id")
```
To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.

jcoux commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -19,11 +19,14 @@ class SaleCoupon(models.Model):
+    discount_fixed_amount_delta = fields.Monetary(
+        "Fixed Amount Delta",
+        compute="_compute_discount_fixed_amount_delta",
+        currency_field="currency_program_id",
```
To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.

oerp-odoo replied:
This does not make any sense. All `sale_coupon_multi_currency` do is make sale_coupon module convert currencies both ways -> from program to SO and from SO to program.

Actual multi-currency capability is already used on sale_coupon module (standard module. As a matter of fact, all Odoo modules handle currencies where monetary amounts are involved without any special extra modules -> its just hidden if multiple currencies are not used), so it must be on sale_coupon_multi_use. It has nothing to do with `sale_coupon_multi_currency`. If I would move all multi-currency awareness to `sale_coupon_multi_currency` (and glue module), then module would work incorrectly if you have multiple currencies, but not install that module. It would just be broken.

And because `sale_coupon_multi_use` introduces part where it tracks discount from SO, it must convert back to his own currency (if its different).

jcoux commented on sale_coupon_multi_use/models/sale_coupon.py:
```diff
@@ -57,9 +60,22 @@ def _get_related_sale_order_line(self, sale_order):
         return {
             "coupon_id": self.id,
-            "amount": abs(sale_order_line.price_total),
+            "amount": amount,
```
To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.

jcoux commented on sale_coupon_multi_use/models/sale_coupon_consumption_line.py:
```diff
@@ -16,7 +16,8 @@ class SaleCouponConsumptionLine(models.Model):
     )
-    amount = fields.Float()
+    currency_program_id = fields.Many2one(related="coupon_id.program_id.currency_id")
+    amount = fields.Monetary(currency_field="currency_program_id")
```
To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.

oerp-odoo replied:
Yes it is linked, it has nothing to do with `sale_coupon_multi_currency`. Also if I would move it there, module would crash, because it does not know anything about sale.coupon.consumption_line model.

jcoux commented on sale_coupon_multi_use/tests/common.py:
```diff
@@ -17,11 +17,34 @@ def setUpClass(cls):
+        cls.company_main = cls.env.ref("base.main_company")
+        cls.eur = cls.env.ref("base.EUR")
+        cls.usd = cls.env.ref("base.USD")
+        cls.pricelist_public = cls.env.ref("product.list0")
```
To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.

jcoux commented on sale_coupon_multi_use/views/sale_coupon_views.xml:
```diff
@@ -24,6 +25,7 @@
                     attrs="{'invisible': [('multi_use', '=', False)]}"
                 >
                     <tree>
+                        <field name="currency_program_id" invisible="1" />
```
To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.

jcoux commented on sale_coupon_multi_use_currency/i18n/fr.po:
```diff
@@ -0,0 +1,27 @@
+#: code:addons/sale_coupon_multi_use_currency/models/sale_coupon_program.py:0
+#, python-format
+msgid "Currency can't be changed when there are Multi Use coupons already."
+msgstr ""
```
```suggestion
msgstr "La devise ne peut être changée quand on a déjà plusieurs utilisations du coupon."
```

jcoux reviewed (changes requested):
Seems we still have multi currency behavior into the multi use module.
I indicate some:
`To move in your `sale_coupon_multi_currency` module: it's not linked to multi use.`
You must be able to install multi currency without multi use.
And the inverse is true too.

oerp-odoo committed:
fixup! [REF] sale_coupon_multi|use|currency

jcoux reviewed (approved):
`sale_coupon_multi_use` and `sale_coupon_multi_currency` have been tested independently by @epanisset and it works!
Then I approve this! :)
